### ISSUE TITLE
Items from const literals should be able to be const as well

### ISSUE THREAD

**Issue by peterscodee** [dart-lang/language#1851]
It would be a convenient feature to be able to declare `const` variables from `const` literals. 

```dart
const List<int> list = [1, 2, 3];
const Map<String, int> map = {"key": 1};

void main() {
  const int fromList1 = list[0]; //should work
  const int fromList2 = list[7]; //should not work
  
  const int fromMap1 = map["key"]; //should work
  const int fromMap2 = map["xyz"]; //should not work
}
```

**Comment by ykmnkmi** [comment]
same for constant object fields.

**Comment by PlugFox** [comment]
Why not just use it like that?

```
const String key = 'key';
const List<String> list = <String>[key];
```

**Comment by Levi-Lesches** [comment]
Related to and maybe duplicate of https://github.com/dart-lang/sdk/issues/3059.

@PlugFox, sometimes you need a bit more control than that: 

```dart
class Person {
  final String name;
  final int age;
  const Person(this.name, this.age);
  const Person.fromJson(Map json) : 
    name = json["name"],  // <-- error, even though json is const
    age = json["age"];
}

void main() {
  const Map json = {"name": "John Smith", "age": 42};
  const Person john = Person.fromJson(json);
}
```

Here I used `Map.[]`, but you can imagine a similar example with `List.[]`.

**Comment by munificent** [comment]
Unfortunately, this isn't safe. Just because you have a const instance, that doesn't mean that every operation you can perform on it can reliably be executed at compile time. This is a valid program:

```dart
class ShadyMap implements Map {
  const ShadyMap(); // Const constructor.

  dynamic operator [](dynamic key) {
    // Definitely not safe to do at compile time!
    if (DateTime.now().hour < 12) return 'morning';
    return 'afternoon';
  }

  // Other Map members...
}

void main() {
  const Map json = ShadyMap();
  const Person john = Person.fromJson(json);
}
```

Dart doesn't currently have any notion of const *methods*, which is what you'd need to be able to determine what's safe to perform.

**Comment by Levi-Lesches** [comment]
I think whats being asked is for the compiler to recognize a `const` Map/Set/List specifically as a special case with a platform implementation that _is_ safe and optimized. Would that be possible?

**Comment by munificent** [comment]
> I think whats being asked is for the compiler to recognize a `const` Map/Set/List specifically as a special case with a platform implementation that _is_ safe and optimized. Would that be possible?

Yeah, in principle, during const evaluation (or attempted const evaluation), we could look at the actual *value* that the method is being called on and if it's one of a known blessed set of concrete types (and not subtypes) then we could treat the method call as a valid const expression. That feels like a fairly fragile boundary to me, and I'm not sure how it would feel in practice.

**Comment by natebosch** [comment]
> We could probably allow index access on a `List` or `Map` literal value.

This would mean that changing a constant in a library from a `dart:core` collection to a sub-type of is a breaking change.

**Comment by lrhn** [comment]
@natebosch Correct. 
So is, currently, changing an object implementing `Symbol` or `Type` from the platform implementation to your own version. Unlikely to be as common, or as useful, as indexing list/maps though.
